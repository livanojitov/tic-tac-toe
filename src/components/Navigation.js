import { NavLink }         from 'react-router-dom';
import React, { useState } from 'react';
import * as constants      from './Constants';
const { HOME, HISTORY, ABOUT, CONTACT } = constants;

const Navigation = () => {

  const [routes] = useState([
    { name: HOME,    url: '/'        },
    { name: HISTORY, url: '/history' },
    { name: ABOUT,   url: '/about'   },
    { name: CONTACT, url: '/contact' }
  ]);

  const lis = routes.map((route, key) => {
    return  (
      <li key={key}><NavLink exact to={route.url}>{route.name}</NavLink></li>
    )  
  });

  return (
    <nav>
      <input type="checkbox" id="menu" />
      <label class="mylabel" for="menu">
        <div id="burger-container">
            <div id="burger">
              <span>&nbsp;</span>
              <span>&nbsp;</span>
              <span>&nbsp;</span>
            </div>
          </div>
      </label>      
      <ul className='nav'>
        {lis}                  
      </ul>
    </nav>
  )
}

export default Navigation;