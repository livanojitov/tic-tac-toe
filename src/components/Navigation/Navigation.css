nav{ 
  display:inline-block;
  width: 70%;  
}

nav ul{
  list-style: none;
  padding-left: 0;
}

nav li{
  display: inline-block;
  margin-right: 10px;
  cursor: pointer;
}

nav a.active{
  font-weight: 1000;
}

a{
  color: green;
  text-decoration: none;  
}

nav li a:hover{
  color: #19A57E
}

input#menu{
  display: none;
}

label.mylabel{
  display: none;
  cursor: pointer;
} 

#burger-container{
  position: relative;
  width: 100%;
}

#burger{
  display: block;
}

#burger span{
  background: green;
  display: block;
  left: 80%;
  width: 30px;
  height: 3px;
  margin-bottom: 6px;
  position: relative;
  top: 0;
  transition: all ease-in-out 0.4s;
}

@media screen and (max-width: 400px) {
  nav{
    width: 60%;
  }

  nav li{
    display: block;
  }

  nav a{
    display: block;
    text-decoration: none;
    padding: 5px;
    background-color: #8fbc8f;
  }

  ul.nav{
    margin: 0;
    display: block;
    width: 100%;
    max-height: 0;
    overflow: hidden;
    text-align: left;
  }

  label.mylabel{
    display: inline-block;
    width: 100%;
    cursor: default;
  }
  
  input#menu:checked ~ ul{
    max-height: 160px;
    transition: all ease 1s;
  }
  
  input#menu:checked + label > #burger-container span:nth-child(2){
    width: 0;
    opacity: 0;
  }
  
  input#menu:checked + label > #burger-container span:nth-child(3){
    transform: rotate(45deg);
    top: -9px;
  }
  
  input#menu:checked + label > #burger-container span:nth-child(1){
    transform: rotate(-45deg);
    top: 9px;
  }

  input#menu ~ ul a.active{
    position: absolute;
    top: 5px;
  }
  input#menu:checked ~ ul a.active{
    position: relative;
    top: 0px;
  }    
}