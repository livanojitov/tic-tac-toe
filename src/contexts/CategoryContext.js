import React, { createContext } from 'react';
import useFetchData from '../hooks/useFetchData';

export const CategoryContext = createContext();

const CategoryContextProvider = ({children}) => {
  let {data: categories, error} = useFetchData("categories");

  return (
    <CategoryContext.Provider value={{categories, error}}>
      {children}
    </CategoryContext.Provider>
  );

}
 
export default CategoryContextProvider;