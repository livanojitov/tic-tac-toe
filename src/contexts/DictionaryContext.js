import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios'

export const DictionaryContext = createContext();

const DictionaryContextProvider = (props) => {
  const [DICTIONARY, setDictionary] = useState('');

  useEffect(() => {
    if (DICTIONARY.length === 0){
      axios.get('dictionary.json')
        .then(res => {
          setDictionary(res.data);
        })
    }
  });

  return (
    <DictionaryContext.Provider value={{ DICTIONARY }}>
      {props.children}
    </DictionaryContext.Provider>
  );

}
 
export default DictionaryContextProvider;