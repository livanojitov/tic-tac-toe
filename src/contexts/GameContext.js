import React, { createContext } from 'react';
import useLocalStorage          from '../hooks/useLocalStorage';
export const GameContext = createContext();

const GameContextProvider = ({children}) => {

  const {games, dispatch} = useLocalStorage('tictactoe', []);

  const getGame = (id) => {
    if (games.length > 0){
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  }

  return (
    <GameContext.Provider value={{games, dispatch, getGame}}>
      {children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;