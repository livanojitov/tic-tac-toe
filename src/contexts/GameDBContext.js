import React, { createContext } from "react";
import useLocalStorage from "../hooks/useLocalStorage";
export const GameDBContext = createContext();

const GameDBContextProvider = ({ children }) => {
  const { games, dispatch } = useLocalStorage("tictactoe", []);

  const getGame = (id) => {
    if (games.length > 0) {
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  };

  return (
    <GameDBContext.Provider value={{ games, dispatch, getGame }}>
      {children}
    </GameDBContext.Provider>
  );
};

export default GameDBContextProvider;
