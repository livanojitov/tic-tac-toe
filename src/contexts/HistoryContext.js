import React, { createContext, useState } from 'react';

export const HistoryContext = createContext();

const HistoryContextProvider = (props) => {

  const [history, setHistory] = useState(0);

  const changeHistory = (history) => {
    setHistory(history);
  }

  return (
    <HistoryContext.Provider value={{ history, changeHistory }}>
      {props.children}
    </HistoryContext.Provider>
  );

}
 
export default HistoryContextProvider;