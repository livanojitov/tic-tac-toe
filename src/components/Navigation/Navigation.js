import { NavLink }            from 'react-router-dom';
import React, { useContext }  from 'react';
import { DictionaryContext }  from '../../contexts/DictionaryContext';
import './Navigation.css';

const Navigation = () => {
  const { dictionary } = useContext(DictionaryContext);

  return dictionary ? (
    <nav>
      <input type="checkbox" id="menu" />
      <label className="mylabel" htmlFor="menu">
        <div id="burger-container">
          <div id="burger">
            <span>&nbsp;</span>
            <span>&nbsp;</span>
            <span>&nbsp;</span>
          </div>
        </div>
      </label> 
      <ul data-testid="nav" className='nav'>
        <li key='0'><NavLink data-testid="game" exact to={'/'}          >{dictionary.GAME}      </NavLink></li>
        <li key='1'><NavLink data-testid="history" exact to={'/history'}   >{dictionary.HISTORY}   </NavLink></li>
        <li key='2'><NavLink data-testid="repository" exact to={'/repository'}>{dictionary.REPOSITORY}</NavLink></li>
        <li key='3'><NavLink data-testid="about" exact to={'/about'}     >{dictionary.ABOUT}     </NavLink></li>
        <li key='4'><NavLink data-testid="contact" exact to={'/contact'}   >{dictionary.CONTACT}   </NavLink></li>
      </ul>
    </nav>
  ) : (null)
}

export default Navigation;