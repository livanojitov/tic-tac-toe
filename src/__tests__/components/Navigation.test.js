import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import Navigation from '../../components/Navigation/Navigation';

describe('Navigation component', () => {
  let dictionary;
  let history;

  beforeEach(() => {
    dictionary = {GAME:       "Game",
                  HISTORY:    "Historical",
                  REPOSITORY: "Repository",
                  ABOUT:      "About",
                  CONTACT:    "Contact"};
  });

  afterEach(cleanup);

  const renderNavigation = () => {
    history = createMemoryHistory();
    history.push = jest.fn();
    
    return render(
      <DictionaryContext.Provider value={{dictionary}}>
        <Router history={history}>
          <Navigation/>
        </Router>
      </DictionaryContext.Provider>
    );
  }

  test("renders the Navigation component correctly", () => {
    const { asFragment } = renderNavigation();
    expect(asFragment()).toMatchSnapshot();
  })

  test("nav", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('nav')).toHaveClass('nav');
  })

  test("game description", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('game')).toHaveTextContent('Game');
  })

  test("historical description", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('history')).toHaveTextContent('Historical');
  })

  test("repository description", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('repository')).toHaveTextContent('Repository');
  })

  test("about description", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('about')).toHaveTextContent('About');
  })

  test("contact description", () => {
    const { getByTestId } = renderNavigation();
    expect(getByTestId('contact')).toHaveTextContent('Contact');
  })

  test("routes to the Game component", () => {
    const { getByTestId } = renderNavigation();
    fireEvent.click(getByTestId('game'));
    expect(history.push).toHaveBeenCalledWith({hash: "", pathname: "/", search: "", state: null});
    expect(getByTestId('game')).toHaveAttribute('href', '/');
  });

  test("routes to the History component", () => {
    const { getByTestId } = renderNavigation();
    fireEvent.click(getByTestId('history'));
    expect(history.push).toHaveBeenCalledWith({hash: "", pathname: "/history", search: "", state: null});
    expect(getByTestId('history')).toHaveAttribute('href', '/history');
  });

  test("routes to the Repository component", () => {
    const { getByTestId } = renderNavigation();
    fireEvent.click(getByTestId('repository'));
    expect(history.push).toHaveBeenCalledWith({hash: "", pathname: "/repository", search: "", state: null});
    expect(getByTestId('repository')).toHaveAttribute('href', '/repository');
  });


  test("routes to the About component", () => {
    const { getByTestId } = renderNavigation();
    fireEvent.click(getByTestId('about'));
    expect(history.push).toHaveBeenCalledWith({hash: "", pathname: "/about", search: "", state: null});
    expect(getByTestId('about')).toHaveAttribute('href', '/about');
  });
  
  test("routes to the Contact component", () => {
    const { getByTestId } = renderNavigation();
    fireEvent.click(getByTestId('contact'));
    expect(history.push).toHaveBeenCalledWith({hash: "", pathname: "/contact", search: "", state: null});
    expect(getByTestId('contact')).toHaveAttribute('href', '/contact');
  });
})