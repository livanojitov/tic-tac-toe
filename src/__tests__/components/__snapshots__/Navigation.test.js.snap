// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component renders the Navigation component correctly 1`] = `
<DocumentFragment>
  <nav>
    <input
      id="menu"
      type="checkbox"
    />
    <label
      class="mylabel"
      for="menu"
    >
      <div
        id="burger-container"
      >
        <div
          id="burger"
        >
          <span>
             
          </span>
          <span>
             
          </span>
          <span>
             
          </span>
        </div>
      </div>
    </label>
    <ul
      class="nav"
      data-testid="nav"
    >
      <li>
        <a
          aria-current="page"
          class="active"
          data-testid="game"
          href="/"
        >
          Game      
        </a>
      </li>
      <li>
        <a
          data-testid="history"
          href="/history"
        >
          Historical   
        </a>
      </li>
      <li>
        <a
          data-testid="repository"
          href="/repository"
        >
          Repository
        </a>
      </li>
      <li>
        <a
          data-testid="about"
          href="/about"
        >
          About     
        </a>
      </li>
      <li>
        <a
          data-testid="contact"
          href="/contact"
        >
          Contact   
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
